File: Linked-Lists/21_Merge_Two_Sorted_Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {

        ListNode* dummy = new ListNode(-1);
        ListNode* curr = dummy;

        ListNode* t1 = list1;
        ListNode* t2 = list2;

        while(t1 != nullptr && t2 != nullptr) {
            if(t1->val < t2->val) {
                curr->next = t1;  
                t1 = t1->next;   
            } else {
                curr->next = t2;  
                t2 = t2->next;   
            }
            curr = curr->next;
        }

        if(t1 != nullptr) {
            curr->next = t1;
        } else if(t2 != nullptr) {
            curr->next = t2;
        }

        ListNode* mergedHead = dummy->next;
        delete dummy;

        return mergedHead;
    }
};

File: Linked-Lists/21_Merge_Two_Sorted_Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        t1 = list1
        t2 = list2

        dummy = ListNode(-1)
        curr = dummy

        while t1 is not None and t2 is not None:
            if t1.val < t2.val:
                curr.next = t1
                t1 = t1.next
            else:
                curr.next = t2
                t2 = t2.next
            
            curr = curr.next
        
        if t1 is not None:
            curr.next = t1
        elif t2 is not None:
            curr.next = t2
        
        return dummy.next
        
